% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={RNA-seq analysis and Shiny application},
  pdfauthor={David GALLIEN \& Gabin COUDRAY},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering

\title{RNA-seq analysis and Shiny application}
\author{David GALLIEN \& Gabin COUDRAY}
\date{15/04/2020}

\begin{document}
\maketitle

~

\hypertarget{introduction}{%
\subsection{Introduction}\label{introduction}}

\textless p style=``text-align:justify'';\textgreater{} Most of
biological researchers does not have time to use the tools which allow
to analyze the data generated by new technology. It is why it is
imperative to offer to biologist facilitate tools that can allow them to
be more efficient and conscenrate on there research.

\textless p style=``text-align:justify'';\textgreater{} In our project
we want to be able to generate application that can help them to explore
there data and their results in R environment. For that, Shiny package
allows to create interactive web applications.

\textless p style=``text-align:justify'';\textgreater{} Nevertheless to
create a vizualisation interactive application we need to have something
to show. In medical research which is the main field we want to explore,
next generation sequencing are often use and generate big data which we
need to be analyzed and explored. We decided to concentrate our work on
RNA sequencing (RNA-seq), this next generation sequencing had the
purpose to detect differential express between cellular type of
different condition.

\hypertarget{objectives}{%
\subsection{Objectives}\label{objectives}}

\hypertarget{context}{%
\subsubsection{Context}\label{context}}

An RNA sequencing analysis presents 3 main steps :

\begin{itemize}
\item
  Random fragmentation of mature RNA
\item
  Amplification of fragments by PCR
\item
  Sequencing of these amplified fragments making millions of reads
\end{itemize}

\hypertarget{rna-sequencing-data-analysis}{%
\subsubsection{RNA-sequencing data
analysis}\label{rna-sequencing-data-analysis}}

\hypertarget{step-1-data-generation}{%
\paragraph{Step 1 : Data generation}\label{step-1-data-generation}}

\hypertarget{step-2-quality}{%
\paragraph{Step 2 : Quality}\label{step-2-quality}}

\hypertarget{step-3-mapping}{%
\paragraph{Step 3 : Mapping}\label{step-3-mapping}}

\hypertarget{step-4-quantification}{%
\paragraph{Step 4 : Quantification}\label{step-4-quantification}}

\hypertarget{step-5-statistics}{%
\paragraph{Step 5 : Statistics}\label{step-5-statistics}}

\hypertarget{material-and-methods}{%
\subsubsection{Material and methods}\label{material-and-methods}}

\hypertarget{deseq2}{%
\subsubsection{DESeq2}\label{deseq2}}

\hypertarget{what-is-deseq2}{%
\paragraph{What is DESeq2 ?}\label{what-is-deseq2}}

\hypertarget{dataset}{%
\paragraph{Dataset}\label{dataset}}

\hypertarget{shiny-application}{%
\subsubsection{Shiny application}\label{shiny-application}}

~

First we will explain more deeply the RNA-seq analysis then we will
proceed to the count table analysis under R thanks to the package
DESeq2. Finally, we will create the Shiny app to study the results of
the analysis.

\end{document}
